opt/constraints: add Implies and CutFirstColumn methods

RaduBerinde opened the pull request:
Implies checks if a constraint is implied by ("included" in) another
one. Also some misc cleanup.

CutFirstColumn cuts out the first column and tries to deduce a
constraint on the remaining columns.

Both primitives are used during filter simplification after index
constraints.

Release note: None

RaduBerinde committed:
opt/constraints: add SubsetOf and CutFirstColumn methods

SubsetOf checks if a constraint is implied by ("included" in) another
one. Also some misc cleanup.

CutFirstColumn cuts out the first column and tries to deduce a
constraint on the remaining columns.

Both primitives are used during filter simplification after index
constraints.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24098)
<!-- Reviewable:end -->

andy-kimball commented:
Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---

*[pkg/sql/opt/constraint/constraint.go, line 247 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L85BFZX-9Ju2kEmXXER:-L85BFZX-9Ju2kEmXXES:b-ft6auj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a985065f3e411781dc66725a878d4d9c6573f70a/pkg/sql/opt/constraint/constraint.go#L247)):*
> ```Go
> 				// There is no span that overlaps with leftSpan.
> 				return false
> 			}
> ```

Shouldn't there be a `continue` here? What happens if there are multiple right spans that start before the first left span? If this turns out to be a problem, make sure to add a test case to regress it.

---

*[pkg/sql/opt/constraint/constraint.go, line 249 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L85B49mECwp_5HYl-Jj:-L85B49mECwp_5HYl-Jk:b-mgcg8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a985065f3e411781dc66725a878d4d9c6573f70a/pkg/sql/opt/constraint/constraint.go#L249)):*
> ```Go
> 			}
> 		}
> 		rightSpan := right.Get(rightIndex)
> ```

Seems you should move this above the `if` statement, since you use it there as well. No reason to get it twice...

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Hm, it seems like I need a more complex primitive. Something that can figure out that a constraint `/a/b: [/1/2 - /1/3]` implies a constraint `/b: [/2 - /3]`. I'll have to rework this.

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/opt/constraint/constraint.go, line 247 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L85BFZX-9Ju2kEmXXER:-L85EZA94uDq6nNmoG37:b-lzwba5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a985065f3e411781dc66725a878d4d9c6573f70a/pkg/sql/opt/constraint/constraint.go#L247)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Shouldn't there be a `continue` here? What happens if there are multiple right spans that start before the first left span? If this turns out to be a problem, make sure to add a test case to regress it.
</blockquote></details>

Ah, that was supposed to be a `for`. I'll add a test.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-:-L85EhLN1bYfFNXqG7WN:b-zcnbb1)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Added a `CutFirstColumn` primitive and related tests.

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 5 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/constraint/constraint.go, line 77 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88i8sV6xa1MdM4BOn-:-L88i8sV6xa1MdM4BOn0:bmh885p) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L77)):*
> ```Go
> // constraints.
> //
> // The two constraints must be on the same set of columns.
> ```

You already said this in the paragraph above -- "The columns of both constraints must be the same"

---

*[pkg/sql/opt/constraint/constraint.go, line 166 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88iJn4EcBBtE4iOcD8:-L88iJn4EcBBtE4iOcD9:b-89a4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L166)):*
> ```Go
> // should be marked as empty and all constraints removed.
> //
> // The two constraints must be on the same set of columns.
> ```

Ditto

---

*[pkg/sql/opt/constraint/constraint.go, line 240 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88iiFLCj5GTyi5zG6R:-L88iiFLCj5GTyi5zG6S:b-thjyr0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L240)):*
> ```Go
> 	keyCtx := MakeKeyContext(&c.Columns, evalCtx)
> 
> 	for leftIndex, rightIndex := 0, 0; leftIndex < left.Count(); leftIndex++ {
> ```

Do you need to check if `rightIndex < right.Count()`? What if right has no spans?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-:-L88pKH2AX0vu0sUfF5b:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 8 of 10 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/constraint/constraint.go, line 77 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88i8sV6xa1MdM4BOn-:-L88rGuUEZbE8VoaWzKC:bd55tvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L77)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

You already said this in the paragraph above -- "The columns of both constraints must be the same"
</blockquote></details>

Ha, indeed, fixed.

---

*[pkg/sql/opt/constraint/constraint.go, line 166 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88iJn4EcBBtE4iOcD8:-L88rJ6M577si4K2QxDa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L166)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint.go, line 240 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88iiFLCj5GTyi5zG6R:-L88sXmf5DJ10e0-izBc:b-ol96g7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L240)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Do you need to check if `rightIndex < right.Count()`? What if right has no spans?
</blockquote></details>

Good catch. Fixed and added a test.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/constraint/constraint_test.go, line 211 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88x7_EA3keiW8sOmEW:-L88x7_EA3keiW8sOmEX:bde1lll) ([raw file](https://github.com/cockroachdb/cockroach/blob/48b4958ea33e7faf38c728268dcb1926828bc821/pkg/sql/opt/constraint/constraint_test.go#L211)):*
> ```Go
> 	test(data.c60to70, right, true)
> 
> 	right = data.c1to10
> ```

Can you add a comment to say what this is testing? It's not immediately obvious.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/constraint/constraint.go, line 231 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88rCvy-kfexlQSD6WD:-L88rCvy-kfexlQSD6WE:bb45la0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L231)):*
> ```Go
> //
> // The two constraints must be on the same set of columns.
> func (c *Constraint) Implies(evalCtx *tree.EvalContext, other *Constraint) bool {
> ```

NIT: Have you considered the name `SubsetOf` or `ContainedIn`?

---

*[pkg/sql/opt/constraint/constraint.go, line 266 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88r_qj9F-Jy5GdjwfF:-L88r_qj9F-Jy5GdjwfG:bvdrvdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L266)):*
> ```Go
> // We can deduce a constraint only if all the spans have equal start and end
> // keys on the first column. For example:
> //   before: /0/1/2/3: [/1/5 - /1/6] [/2/3 - /2/4]
> ```

NIT: I'd use `/a/b/c/d` instead of `/0/1/2/3`, to eliminate any confusion between columns vs. values.

---

*[pkg/sql/opt/constraint/constraint.go, line 272 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88sCJc9DJSIMHL5O5h:-L88sCJc9DJSIMHL5O5i:b-bobm9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L272)):*
> ```Go
> //   after: /1/2/3: unconstrained.
> func (c *Constraint) CutFirstColumn(evalCtx *tree.EvalContext) {
> 	keyCtx := MakeKeyContext(&c.Columns, evalCtx)
> ```

Don't you need immutability check here? What would stop caller from calling `CutFirstColumn` on a constraint that was expected to be immutable? `Constraint` is the only struct that doesn't protect itself in some way (i.e. have immutability bit or always make copies).

---

*[pkg/sql/opt/constraint/constraint.go, line 281 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88skG8Ea0SlWjilBCu:-L88skG8Ea0SlWjilBCv:bz3pzr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L281)):*
> ```Go
> 	for i := 0; i < c.Spans.Count(); i++ {
> 		sp := c.Spans.Get(i)
> 		if sp.start.Length() == 0 || sp.end.Length() == 0 ||
> ```

NIT: a bit more efficient and clear would be `sp.start.IsEmpty()  || sp.end.IsEmpty()`. Or maybe a helper method on `Key`, maybe `func (k Key) HasEqualValues(nth int)`?

---

*[pkg/sql/opt/constraint/spans.go, line 119 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88umdp911KnF4KfGlM:-L88umdp911KnF4KfGlN:b-ewaitz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/spans.go#L119)):*
> ```Go
> // sortedAndNonOverlapping returns true if the collection of spans is strictly
> // ordered and no spans overlap.
> func (s *Spans) sortedAndNonOverlapping(keyCtx *KeyContext) bool {
> ```

NIT: Maybe just `sortedAndMerged`. Also consider a bool in `Spans` that gets set once `SortAndMerge` is called so that no scan has to happen (but only if it would help real scenarios).

---

*[pkg/sql/opt/constraint/spans.go, line 129 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88tofS-em_l-oy0O0h:-L88tofS-em_l-oy0O0i:by8uhdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/spans.go#L129)):*
> ```Go
> 
> // SortAndMergeOverlaps sorts the spans and merges any overlapping spans.
> func (s *Spans) SortAndMergeOverlaps(keyCtx *KeyContext) {
> ```

NIT: I think just `SortAndMerge` is fine, since I think the word "Merge" implies overlaps, as what else could be merged except overlapping spans?

---

*[pkg/sql/opt/constraint/spans.go, line 160 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88yXGIE5mYogCb5A2E:-L88yXGIE5mYogCb5A2F:bwhx654) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/spans.go#L160)):*
> ```Go
> // SortAndDedup sorts the spans (according to Span.Compare) and removes any
> // duplicates.
> func (s *Spans) SortAndDedup(keyCtx *KeyContext) {
> ```

Are you sure you need this method as well as `SortAndMerge`? Can't you just callers always use `SortAndMerge`, since it's not more expensive than `SortAndDedup` in the case where there are no overlapping spans.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 6 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/opt/constraint/constraint.go, line 231 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88rCvy-kfexlQSD6WD:-L894YHCAxm2hFES7j6_:brh9q63) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L231)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: Have you considered the name `SubsetOf` or `ContainedIn`?
</blockquote></details>

Changed to `SubsetOf`

---

*[pkg/sql/opt/constraint/constraint.go, line 266 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88r_qj9F-Jy5GdjwfF:-L894fxQ69nHW-vjBU2Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L266)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: I'd use `/a/b/c/d` instead of `/0/1/2/3`, to eliminate any confusion between columns vs. values.
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint.go, line 272 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88sCJc9DJSIMHL5O5h:-L8931J3DrR6ULrb9H17:b-tesyro) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L272)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Don't you need immutability check here? What would stop caller from calling `CutFirstColumn` on a constraint that was expected to be immutable? `Constraint` is the only struct that doesn't protect itself in some way (i.e. have immutability bit or always make copies).
</blockquote></details>

Pretty much all operations with a Constraint modify the receiver. The ConstraintSet stores Constraints by value. The only thing we need to ensure is that we don't inadvertently change the spans of a copy of this `Constraint` (which is enforced by the immutable flag on Spans).

---

*[pkg/sql/opt/constraint/constraint.go, line 281 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88skG8Ea0SlWjilBCu:-L894usX2CANrQrnW9qH:b-k9d12p) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/constraint.go#L281)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: a bit more efficient and clear would be `sp.start.IsEmpty()  || sp.end.IsEmpty()`. Or maybe a helper method on `Key`, maybe `func (k Key) HasEqualValues(nth int)`?
</blockquote></details>

Changed to IsEmpty

---

*[pkg/sql/opt/constraint/constraint_test.go, line 211 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88x7_EA3keiW8sOmEW:-L8958EP25H1csQggKlL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/48b4958ea33e7faf38c728268dcb1926828bc821/pkg/sql/opt/constraint/constraint_test.go#L211)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Can you add a comment to say what this is testing? It's not immediately obvious.
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/spans.go, line 119 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88umdp911KnF4KfGlM:-L895No46W0oUKPANPt4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/spans.go#L119)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: Maybe just `sortedAndMerged`. Also consider a bool in `Spans` that gets set once `SortAndMerge` is called so that no scan has to happen (but only if it would help real scenarios).
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/spans.go, line 129 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88tofS-em_l-oy0O0h:-L895OZJElIHHlp5YOHy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/spans.go#L129)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: I think just `SortAndMerge` is fine, since I think the word "Merge" implies overlaps, as what else could be merged except overlapping spans?
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/spans.go, line 160 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L88yXGIE5mYogCb5A2E:-L894Ad4Cvd26Wn1GrlG:bbbecmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1677b0309c792356a12c3048e32bbe7931a2dcb8/pkg/sql/opt/constraint/spans.go#L160)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Are you sure you need this method as well as `SortAndMerge`? Can't you just callers always use `SortAndMerge`, since it's not more expensive than `SortAndDedup` in the case where there are no overlapping spans.
</blockquote></details>

Good point, removed SortAndDedup.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 6 of 10 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 176 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L89Jnm17wjp9bEcxWbd:-L89Jnm17wjp9bEcxWbe:b-ylav2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/24c37398dd9a083465c273410ac348780323020b/pkg/sql/opt/constraint/constraint_set.go#L176)):*
> ```Go
> // example:
> //   (x > 1 AND y > 10) OR (x < 5 AND y < 50)
> // the union is unconstrained (and thus allows combinations like x,y = 2,0).
> ```

NIT: maybe a slightly better example would be `x, y = 10, 0`, since that combo would be rejected by both sides, which links back to "neither of the input sets allowed".

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-:-L89KYufDH2MUbX_lhyP:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 5 of 10 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 176 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098#-L89Jnm17wjp9bEcxWbd:-L89Ls5oCUcLt_VHcu0e:b124xv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/24c37398dd9a083465c273410ac348780323020b/pkg/sql/opt/constraint/constraint_set.go#L176)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: maybe a slightly better example would be `x, y = 10, 0`, since that combo would be rejected by both sides, which links back to "neither of the input sets allowed".
</blockquote></details>

Thanks, that was my intention, fixed.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24098)*
<!-- Sent from Reviewable.io -->
